luigi/contrib/pai.py

```diff
@@ -0,0 +1,300 @@
+              "minFailedTaskCount": Integer,
+              "minSucceededTaskCount": Integer
+            }
+          ],
+          "gpuType": String,
+          "retryCount": Integer
+        }
+
+    """
+    __slots__ = (
+        'jobName', 'image', 'authFile', 'dataDir', 'outputDir', 'codeDir', 'virtualCluster',
+        'taskRoles', 'gpuType', 'retryCount'
+    )
+
+    def __init__(self, jobName, image, tasks):
+        self.jobName = jobName
+        self.image = image
+        if isinstance(tasks, list) and len(tasks) != 0:
+            self.taskRoles = tasks
+        else:
+            raise TypeError('you must specify one task at least.')
+
+
+class Port(object):
+    __slots__ = ('label', 'beginAt', 'portNumber')
+
+    def __init__(self, label, begin_at=0, port_number=1):
+        """
+        The Port definition for TaskRole
+
+        :param label: Label name for the port type, required
+        :param begin_at: The port to begin with in the port type, 0 for random selection, required
+        :param port_number: Number of ports for the specific type, required
+        """
+        self.label = label
+        self.beginAt = begin_at
+        self.portNumber = port_number
+
+
+class TaskRole(object):
+    __slots__ = (
+        'name', 'taskNumber', 'cpuNumber', 'memoryMB', 'shmMB', 'gpuNumber', 'portList', 'command',
+        'minFailedTaskCount', 'minSucceededTaskCount'
+    )
+
+    def __init__(self, name, command, taskNumber=1, cpuNumber=1, memoryMB=2048, shmMB=64, gpuNumber=0, portList=[]):
+        """
+        The TaskRole of PAI
+
+        :param name: Name for the task role, need to be unique with other roles, required
+        :param command: Executable command for tasks in the task role, can not be empty, required
+        :param taskNumber: Number of tasks for the task role, no less than 1, required
+        :param cpuNumber: CPU number for one task in the task role, no less than 1, required
+        :param shmMB: Shared memory for one task in the task role, no more than memory size, required
+        :param memoryMB: Memory for one task in the task role, no less than 100, required
+        :param gpuNumber: GPU number for one task in the task role, no less than 0, required
+        :param portList: List of portType to use, optional
+        """
+        self.name = name
+        self.command = command
+        self.taskNumber = taskNumber
+        self.cpuNumber = cpuNumber
+        self.memoryMB = memoryMB
+        self.shmMB = shmMB
+        self.gpuNumber = gpuNumber
+        self.portList = portList
+
+
+class OpenPai(luigi.Config):
+    pai_url = luigi.Parameter(
+        default='http://127.0.0.1:9186',
+        description='rest server url, default is http://127.0.0.1:9186')
+    username = luigi.Parameter(
+        default='admin',
+        description='your username')
+    password = luigi.Parameter(
+        default=None,
+        description='your password')
+    expiration = luigi.Parameter(
```

Comment:
Is this expiration really expected as a string? Not an int?

Reply:
It can accept both string and integer, the integer will be better.